tests/test_cli.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+import asyncio
+import os
+import pathlib
+import shutil
+import subprocess
+from typing import Optional
+
+import pytest
+import requests
+import sqlalchemy as sa
+import testing.postgresql
+
+from explog.format_http_request import format_http_request
+
+# Time limit for `explog create-table` (sec).
+CREATE_TIMEOUT = 5
+
+# Time limit for terminating a process (sec).
+TERMINATE_TIMEOUT = 5
+
+# Time to pause after executing `explog run` before trying to use it (sec).
+RUN_DELAY = 2
+
+
+@pytest.mark.asyncio
+async def test_cli() -> None:
+    """Test explog create-table and run command-line commands."""
+    repo_path = pathlib.Path(__file__).parent / "data" / "hsc_raw"
+    os.environ["BUTLER_URI_1"] = str(repo_path)
+
+    exe_path = shutil.which("explog")
+    assert (
+        exe_path is not None
+    ), "Could not find 'explog' bin script; you must build this package"
+
+    with testing.postgresql.Postgresql() as postgresql:
+        os.environ["EXPOSURE_LOG_DATABASE_URL"] = postgresql.url()
+        engine = sa.create_engine(postgresql.url())
+
+        create_table_process = await asyncio.create_subprocess_exec(
+            "explog",
+            "create-table",
+            stderr=subprocess.PIPE,
+        )
+
+        await asyncio.wait_for(
+            create_table_process.wait(), timeout=CREATE_TIMEOUT
+        )
+        data = await create_table_process.stderr.read()  # type: ignore
+        assert (
+            create_table_process.returncode == 0
+        ), f"explog create-table failed with {data.decode()!r}"
+
+        table_names = sa.inspect(engine).get_table_names()
+        assert "exposure_log_messages" in table_names
+
+        for port, message_id in ((None, 1), (8000, 2)):
+            await check_run_and_add_message(port=port, message_id=message_id)
+
+
+async def check_run_and_add_message(
+    port: Optional[int], message_id: int
+) -> None:
+    """Run `explog run` and use it to add one message.
+
+    Before calling this you must have a database running.
+
+    Parameters
+    ----------
+    port
+        Port on which to run the explog service.
+    message_id
+        Expected ID of the added message.
+    """
+
+    cmdline_args = ["explog", "run"]
+    if port is None:
+        port = 8080
```

Comment:
I'm curious about this.  Why not just put 8080 in the test case above?  I'm also curious why you have two ports.  Presumably one is postgres, what is the other?

Reply:
I am testing specifying a port and not specifying a port -- in other words, whether `explog run` and `explog run --port <port>` both work. 8080 is the default port when using `explog run`.

8000 could be any value. I probably should have picked something that looks less like 8080. It happens to be the default port when running using `tox -e run`. I have no idea how the template generator came up with 8080 and 8000 came from, nor why they differ between the two ways of running the service.

Reply:
I'll improve the doc string and add a code comment to explain what's going on.